### PR TITLE
GPC-506: Authorizer working with Cognito

### PR DESCRIPTION
**Pull request by hannah-leong**

### CODE REVIEW

File: lambdas/authorisation/src/test/java/uk/gov/di/data/lep/AuthorisationTest.java [hunk 1]
```diff
@@ -4,28 +4,50 @@
...
+    @BeforeEach
+    void refreshSetup() {
+        clearInvocations(config);
+        clearInvocations(context);
+        clearInvocations(decodedJwt);
+        clearInvocations(jwtService);
+    }
 
     @Test
     void authorisationReturnsAnAuthorisorResponseWithCorrectActionAndEffect() {
```
Comment (oskarwilliams) [inline lambdas/authorisation/src/test/java/uk/gov/di/data/lep/AuthorisationTest.java:37]: could we add a test for if the verifier throws an exception

File: lambdas/authorisation/src/main/java/uk/gov/di/data/lep/services/AwsCognitoRSAKeyProvider.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import java.net.URL;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.interfaces.RSAPublicKey;
+
+public class AwsCognitoRSAKeyProvider implements RSAKeyProvider {
+    private final URL awsKidStoreUrl;
+    private final JwkProvider provider;
+
+    public AwsCognitoRSAKeyProvider(String awsCognitoRegion, String awsUserPoolsId) {
+        String url = String.format(
```
Comment (oskarwilliams) [inline lambdas/authorisation/src/main/java/uk/gov/di/data/lep/services/AwsCognitoRSAKeyProvider.java:18]:
```suggestion
        var url = String.format(
```

File: lambdas/authorisation/src/main/java/uk/gov/di/data/lep/services/AwsCognitoRSAKeyProvider.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        } catch (MalformedURLException e) {
+            throw new RuntimeException(String.format("Invalid URL provided, URL=%s", url));
+        }
+        provider = new JwkProviderBuilder(awsKidStoreUrl).build();
+    }
+
+    @Override
+    public RSAPublicKey getPublicKeyById(String kid) {
+        try {
+            return (RSAPublicKey) provider.get(kid).getPublicKey();
```
Comment (oskarwilliams) [inline lambdas/authorisation/src/main/java/uk/gov/di/data/lep/services/AwsCognitoRSAKeyProvider.java:34]: Do we need to cast this? (the answer may be yes)

**Review by oskarwilliams — CHANGES_REQUESTED** [review]

File: lambdas/authorisation/src/main/java/uk/gov/di/data/lep/services/AwsCognitoRSAKeyProvider.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            throw new RuntimeException(String.format("Invalid URL provided, URL=%s", url));
+        }
+        provider = new JwkProviderBuilder(awsKidStoreUrl).build();
+        logger.info(provider);
+    }
+
+    @Override
+    public RSAPublicKey getPublicKeyById(String kid) {
+        try {
+            logger.info(provider.get(kid));
```
Comment (oskarwilliams) [inline lambdas/authorisation/src/main/java/uk/gov/di/data/lep/services/AwsCognitoRSAKeyProvider.java:38]: Do we want to be doing all this logging? I think not

File: lambdas/authorisation/src/test/java/uk/gov/di/data/lep/AuthorisationTest.java [hunk 5]
```diff
@@ -4,32 +4,70 @@
...
+        when(jwtService.decode(anyString())).thenThrow(new JWTVerificationException("Exception message"));
+
+        var event = new APIGatewayProxyRequestEvent()
+            .withHeaders(
+                Map.of(
+                    "Authorization",
+                    "Bearer accessToken"))
+            .withRequestContext(new ProxyRequestContext().withIdentity(new RequestIdentity()));
+
+        assertThrows(RuntimeException.class, () -> underTest.handleRequest(event, context));
```
Comment (oskarwilliams) [inline lambdas/authorisation/src/test/java/uk/gov/di/data/lep/AuthorisationTest.java:70]: Could we assert some details of the throw please, you should be able to verify the inner exception of it

**Review by oskarwilliams — CHANGES_REQUESTED** [review]

File: lambdas/authorisation/src/main/java/uk/gov/di/data/lep/services/AwsCognitoRSAKeyProvider.java (hunk 4 above)
Comment (hannah-leong) [reply lambdas/authorisation/src/main/java/uk/gov/di/data/lep/services/AwsCognitoRSAKeyProvider.java:38]: dammit

**Review by oskarwilliams — APPROVED** [review]

**Pull request merged**